pkg/blobinfocache/memory/memory.go

```diff
@@ -120,25 +120,45 @@ func (mem *cache) RecordDigestCompressorName(blobDigest digest.Digest, compresso
 	mem.compressors[blobDigest] = compressorName
 }
 
-// appendReplacementCandidates creates prioritize.CandidateWithTime values for (transport, scope, digest), and returns the result of appending them to candidates.
-func (mem *cache) appendReplacementCandidates(candidates []prioritize.CandidateWithTime, transport types.ImageTransport, scope types.BICTransportScope, digest digest.Digest, requireCompressionInfo bool) []prioritize.CandidateWithTime {
-	locations := mem.knownLocations[locationKey{transport: transport.Name(), scope: scope, blobDigest: digest}] // nil if not present
-	for l, t := range locations {
-		compressorName, compressorKnown := mem.compressors[digest]
-		if !compressorKnown {
-			if requireCompressionInfo {
-				continue
-			}
-			compressorName = blobinfocache.UnknownCompression
+// appendReplacementCandidates creates prioritize.CandidateWithTime values for digest in scopeBucket
+// (or with an unknown location if scope is nil and v2Output is set) with corresponding compression
+// info from compressionBucket (if compressionBucket is not nil), and returns the result of appending them
+// to candidates.
+// v2Output ensures that candidates with unknown location are also appended to the final result ignoring the
+// the field LastSeen.
+func (mem *cache) appendReplacementCandidates(candidates []prioritize.CandidateWithTime, transport types.ImageTransport, scope *types.BICTransportScope, digest digest.Digest, v2Output bool) []prioritize.CandidateWithTime {
+	compressorName := blobinfocache.UnknownCompression
+	compressorNameValue, compressorKnown := mem.compressors[digest]
+	if compressorKnown {
+		compressorName = compressorNameValue
+	}
+	if compressorName == blobinfocache.UnknownCompression && v2Output {
+		return candidates
+	}
+	if scope != nil {
```

Comment:
https://github.com/containers/image/pull/1645#discussion_r967328323 : this should be something like
```go
locations := …
if len(locations) > 0 {
    add entries with locations
} else if v2output {
    add an entry without location
}
```

and then make `scope` a non-pointer again.